File: sdks/node/src/resources/analytics.ts
/**
 * HookSniff SDK — Analytics Resource
 */

import { HttpMethod, HookSniffRequest, type HookSniffRequestContext } from "../request";
import type {
  StatsResponse,
  AnalyticsTrendResponse,
  SuccessRateResponse,
  LatencyTrendResponse,
  DeliveryTrendResponse,
} from "../models";

export interface AnalyticsOptions {
  range?: "1h" | "24h" | "7d" | "30d" | "90d";
  endpoint_id?: string;
}

export class Analytics {
  constructor(private readonly requestCtx: HookSniffRequestContext) {}

  /** Get overall stats. */
  public stats(options?: AnalyticsOptions): Promise<StatsResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/stats");
    request.setQueryParams({ range: options?.range });
    return request.send(this.requestCtx, (json) => json as StatsResponse);
  }

  /** Get delivery trends over time. */
  public deliveryTrends(options?: AnalyticsOptions): Promise<DeliveryTrendResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/analytics/deliveries");
    request.setQueryParams({
      range: options?.range,
      endpoint_id: options?.endpoint_id,
    });
    return request.send(this.requestCtx, (json) => json as DeliveryTrendResponse);
  }

  /** Get success rate analytics. */
  public successRate(options?: AnalyticsOptions): Promise<SuccessRateResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/analytics/success-rate");
    request.setQueryParams({
      range: options?.range,
      endpoint_id: options?.endpoint_id,
    });
    return request.send(this.requestCtx, (json) => json as SuccessRateResponse);
  }

  /** Get latency trends. */
  public latency(options?: AnalyticsOptions): Promise<LatencyTrendResponse> {
    const request = new HookSniffRequest(HttpMethod.GET, "/v1/analytics/latency");
    request.setQueryParams({
      range: options?.range,
      endpoint_id: options?.endpoint_id,
    });
    return request.send(this.requestCtx, (json) => json as LatencyTrendResponse);
  }
}

File: sdks/rust/src/resources/analytics.rs
//! HookSniff API Resource: Analytics
//!
//! Delivery analytics — success rate, latency, delivery stats.

use crate::request::{HookSniffRequest, HookSniffRequestContext, HttpMethod};
use serde::{Deserialize, Serialize};

#[derive(Debug, Serialize, Deserialize)]
pub struct DeliveryStats {
    pub total: u64,
    pub success: u64,
    pub failed: u64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub pending: Option<u64>,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct SuccessRateOutput {
    pub rate: f64,
    pub period: String,
}

#[derive(Debug, Serialize, Deserialize)]
pub struct LatencyOutput {
    pub avg_ms: f64,
    pub p50_ms: f64,
    pub p95_ms: f64,
    pub p99_ms: f64,
    pub period: String,
}

pub struct Analytics {
    ctx: HookSniffRequestContext,
}

impl Analytics {
    pub fn new(ctx: HookSniffRequestContext) -> Self {
        Self { ctx }
    }

    /// Get delivery statistics
    pub fn deliveries(&self) -> Result<DeliveryStats, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Get, "/v1/analytics/deliveries");
        req.send(&self.ctx)
    }

    /// Get success rate
    pub fn success_rate(&self) -> Result<SuccessRateOutput, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Get, "/v1/analytics/success-rate");
        req.send(&self.ctx)
    }

    /// Get latency statistics
    pub fn latency(&self) -> Result<LatencyOutput, Box<dyn std::error::Error>> {
        let req = HookSniffRequest::new(HttpMethod::Get, "/v1/analytics/latency");
        req.send(&self.ctx)
    }
}
